eos-boot-upgrade

```diff
@@ -0,0 +1,278 @@
+
+
+def get_disk():
+    '''Get the booting disk'''
+    disk = Disk.get_booting_device()
+    print('[*] Booting disk: ' + disk.device)
+
+    if disk.label != 'gpt':
+        raise BootUpgradeException('[!] Partition table is not GPT')
+
+    root_partition = disk.get_partition_by_guid('4F68BCE3-E8CD-4DB1-96E7-FBCAF984B709')
+    if not root_partition:
+        raise BootUpgradeException('[!] root partition not found (or wrong GUID)')
+
+    print('[*] Root partition: ' + root_partition.device)
+    return disk
+
+
+def get_esp(disk):
+    '''Get the ESP partition data'''
+    esp_partition = disk.get_partition_by_guid('C12A7328-F81F-11D2-BA4B-00A0C93EC93B')
+    if not esp_partition:
+        raise BootUpgradeException('[!] ESP partition not found (or wrong GUID)')
+    if not esp_partition.device.endswith('1'):
+        raise BootUpgradeException('[!] ESP is not the first partition')
+
+    print('[*] ESP partition: ' + esp_partition.device)
+    return esp_partition
+
+
+def get_dir_size(dir_path):
+    '''Get the directory size'''
+    total_size = 0
+    for (dirpath, _, filenames) in os.walk(dir_path):
+        for f in filenames:
+            fp = os.path.join(dirpath, f)
+            total_size += os.path.getsize(fp)
+    return total_size
+
+
+def check_esp_space(esp_mount, ostree_boot_efi_path):
+    '''Check that we have enough space in the ESP to copy the new data'''
+    ostree_boot_efi_size = get_dir_size(ostree_boot_efi_path)
+    esp_free_space = shutil.disk_usage(esp_mount).free
+    return esp_free_space > ostree_boot_efi_size
+
+
+def dir_cmp(dcmp):
+    '''Helper for directories comparison'''
+    if dcmp.diff_files or dcmp.right_only:
+        return True
+
+    for sub_dcmp in dcmp.subdirs.values():
+        if dir_cmp(sub_dcmp):
+            return True
+    return False
+
+
+def check_for_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Recursively compare the files to upgrade'''
+    dcmp = filecmp.dircmp(esp_mount, ostree_boot_efi_path)
+    return dir_cmp(dcmp)
+
+
+def do_uefi_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Actually start the UEFI upgrade copying the files'''
+    esp_mount_tmp = os.path.join(esp_mount, 'EFI.new')
+
+    print('[*] Updated:')
+    try:
+        file_list = distutils.dir_util.copy_tree(ostree_boot_efi_path,
+                                                 esp_mount_tmp)
+        os.sync()
+
+        for f_source in file_list:
+            f_rp = os.path.relpath(f_source, esp_mount_tmp)
+            f_dest = os.path.join(esp_mount, f_rp)
+
+            if not os.path.isfile(f_dest) or not filecmp.cmp(f_source, f_dest):
+                shutil.move(f_source, f_dest)
+                print(f_rp)
+    finally:
+        if os.path.exists(esp_mount_tmp):
+            distutils.dir_util.remove_tree(esp_mount_tmp)
+
+
+def start_uefi_upgrade():
+    '''Start the UEFI files upgrade'''
+    esp_mount = '/boot/efi'
+    ostree_boot_efi_path = os.path.join(cmdline['ostree'], 'boot/efi')
```

Comment:
We should make the ostree deployment path into a command-line parameter. This would allow us to run this script from eos-updater after deploying a new OS update but *before* rebooting into it. (If we'd had this already, we could have fixed the stuck-on-the-old-deployment bug with an OS update.)